### ISSUE TITLE
[BUG] helm-install-rke2-snapshot-controller jobs are in crashloop backoff post etcd snapshot/restore on an rke2 cluster 

### ISSUE THREAD

**Issue by anupama2501** [rancher/rancher#41174]
**Rancher Server Setup**
- Rancher version: `v2.7.2`
- Installation option (Docker install/Helm Chart): `Docker install`
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): k3s `v1.25.5+k3s1`
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: `v1.24.11+rke2r1` upgraded to `v1.25.7+rke2r1`
- Cluster Type (Local/Downstream): Downstream Infrastructure provider RKE2
 
**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: Standard user
   
**Describe the bug**
etcd snapshot ws taken on cluster with k8s 1.24.11+rke2r1 and then the cluster was upgraded to v1.25. Now if we restore the cluster back to v1.24.11 by using the etcd snapshot, noticing the jobs for the rke2-snapshot go into crashloop backoff and never finish. 

**To Reproduce**
1. Create a rancher server on v2.7.2
2. Create an rke2 cluster on k8s `v1.24.11+rke2r1` and set the snapshot to be saved on s3 by providing the config for s3 on the cluster
3. Take an etcd snapshot and make sure the snapshot gets saved in the s3 bucket
4. Upgrade the cluster to k8s `v1.25.7+rke2r1`
5. Restore the cluster using the snapshot taken in step3
6. Cluster comes back up active.

**Result**
Notice the jobs `helm-install-rke2-snapshot-controller`, `helm-install-rke2-snapshot-controller-crd` and `helm-install-rke2-snapshot-validation-webhook` are in crashloop backoff and following are the error logs 
```
helm_v3 install --set-string global.clusterCIDR=10.42.0.0/16 --set-string global.clusterCIDRv4=10.42.0.0/16 --set-string global.clusterDNS=10.43.0.10 --set-string global.clusterDomain=cluster.local --set-string global.rke2Data
Dir=/var/lib/rancher/rke2 --set-string global.serviceCIDR=10.43.0.0/16 rke2-snapshot-controller /tmp/rke2-snapshot-controller.tgz
2023-04-17T13:17:02.220174749Z Error: INSTALLATION FAILED: chart requires kubeVersion: >= v1.25.7 which is incompatible with Kubernetes v1.24.11+rke2r1
```
**Expected Result**
Expected the job to complete without any errors.

**Additional context**
Not seen on an rke1 cluster. 
Not seen on a k3s cluster also.
SURE-6265

**Comment by brandond** [comment]
The `ke2-snapshot-controller` and `rke2-snapshot-controller-crd` manifests are only packaged with 1.25+. When you rolled back the cluster version to 1.24, the charts for the other packaged components were overwritten with their 1.24 versions. Because there is no 1.24 version of these charts, the 1.25 versions (that are not compatible with 1.24) were left on disk.

RKE2 itself does not support, nor does it have any logic to handle, Kubernetes version downgrades. If Rancher is going to support this, it will need to have logic in it to delete `/var/lib/rancher/rke2/server/manifests/rke2-*` as part of the version downgrade process, to ensure that there aren't any "new" charts left behind that will not be overwritten by the old release.

**Comment by Oats87** [comment]
@brandond while it is relatively trivial to add an instruction that `fmt.Sprintf("rm -rf /var/lib/rancher/%s/server/manifests/%s-*", distro, distro)`, I am slightly concerned about potentially introducing a regression from users that may have their own manifests in that directory that are prefixed with `k3s-` or `rke2`. Additionally, inspecting this directory on a K3s node nets a result like:

```
[root@ck-c7-4 manifests]# ls -l
total 24
-rw-------. 1 root root 1914 Apr 18 17:12 ccm.yaml
-rw-------. 1 root root 4868 Apr 18 17:12 coredns.yaml
-rw-------. 1 root root 3764 Apr 18 17:12 local-storage.yaml
drwx------. 2 root root  227 Apr 18 17:12 metrics-server
-rw-------. 1 root root 1039 Apr 18 17:12 rolebindings.yaml
-rw-------. 1 root root 1072 Apr 18 17:12 traefik.yaml
``` 

whereas on an RKE2 system, I see:
```
[root@ck-c7-2 manifests]# ls -l
total 312
drwxr-xr-x. 2 root root    130 Apr 17 17:37 rancher
-rw-r--r--. 1 root root 133466 Apr 18 15:11 rke2-calico-crd.yaml
-rw-r--r--. 1 root root  16466 Apr 18 15:11 rke2-calico.yaml
-rw-r--r--. 1 root root  25668 Apr 18 15:11 rke2-coredns.yaml
-rw-r--r--. 1 root root  57130 Apr 18 15:11 rke2-ingress-nginx.yaml
-rw-r--r--. 1 root root  13626 Apr 18 15:11 rke2-metrics-server.yaml
-rw-r--r--. 1 root root  14492 Apr 18 15:11 rke2-snapshot-controller-crd.yaml
-rw-r--r--. 1 root root  18656 Apr 18 15:11 rke2-snapshot-controller.yaml
-rw-r--r--. 1 root root  20808 Apr 18 15:11 rke2-snapshot-validation-webhook.yaml
```

This leads me to believe we can potentially hit the same problem in K3s if a new manifest is added but doesn't have the `k3s-*` prefix. 

Are there any unique identifiers that would indicate that the manifests were system specific? I'm not really trying to get into the business of parsing yaml on the fly, but we theoretically should be able to invoke `kubectl get -f <manifest>` if necessary to validate that we should delete the manifest file if it belongs to K3s/RKE2.

**Comment by brandond** [comment]
for RKE2 you can look at `/var/lib/rancher/rke2/data/<VERSION>/charts/` to figure out which manifests are owned by a particular release. Files are staged into to that directory when extracted from the runtime image, and then copied into the manifests dir.

If we'd been more intentional about keeping the `--disable` flag in sync with packaged components on RKE2 I would suggest trying to parse that, but it doesn't contain everything that we're packaging - just things that a user might want to disable manually, not things that are disabled internally based on logic like CNI or cloud provider selection.

**Comment by Oats87** [comment]
I suppose we could use the `current` symlink... at time of etcd restoration/rollback we would be running the shutdown/doing the manifest cleanup, we should never be delivering a plan that "installs" the old version of RKE2.

**Comment by Oats87** [comment]
of course on RKE2 there is no `current` symlink... but on `k3s` there is

**Comment by brandond** [comment]
Yeah, the whole content extraction code path is wildly different between the two products.

**Comment by markusewalker** [comment]
**QA TEST PLAN**

**Scenarios**
|Scenario|Test Case|
|---|---|
|1|As a standard user, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, take an etcd snapshot, upgrade to `v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|
|2|As a standard user, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, take an etcd snapshot using S3, `upgrade to v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|
|3|Setup Rancher `v2.7.2`, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, upgrade Rancher to `v2.7-head`, take an etcd snapshot,  `upgrade to v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|
|4|Setup Rancher `v2.7.2`, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, upgrade Rancher to `v2.7-head`, take an etcd snapshot using S3,  `upgrade to v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|

**Comment by markusewalker** [comment]
Reproduced the issue on `v2.7.2`. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Docker
- Rancher version: `v2.7.2`
- Browser: Chrome

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|As a standard user, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, take an etcd snapshot, upgrade to `v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|:x:|

**REPRO STEPS**

**Scenario 1**
1. As a standard user, provisioned a downstream RKE2 `v1.24.13+rke2r1` node driver cluster.
     - Validated that the cluster and nodes came up as `Active`.
2. Took an etcd snapshot of the cluster.
3. Upgraded the cluster to `v1.25.9+rke2r1`; validated that the cluster and nodes are `Active` and have had their K8s version upgraded.
4. Restored the snapshot with option cluster configuration, Kubernetes version and etcd.
     - Validated that the cluster and nodes are back at `v1.24.13+rke2r1` and they are `Active`.
5. Navigated to top-left menu > cluster > Workloads > Pods and reproduced the issue as shown below:
![image](https://github.com/rancher/rancher/assets/52591819/5979ec62-0e62-43f6-8e40-56d692f14922)

Now that I have reproduced the issue, I will attempt to validate that it is addressed in the latest `v2.7-head`.

**Comment by markusewalker** [comment]
Will retest this once the latest commit with this fix is passing drone.

**Comment by markusewalker** [comment]
Validated that this is addressed in `v2.7-head` https://github.com/rancher/rancher/commit/6ce75a0b57ad9e88cf135254a9b3b40082788eb6. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Docker
- Rancher version: `v2.7-head` https://github.com/rancher/rancher/commit/6ce75a0b57ad9e88cf135254a9b3b40082788eb6
- Browser: Chrome

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|As a standard user, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, take an etcd snapshot, upgrade to `v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|:white_check_mark:|
|2|As a standard user, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, take an etcd snapshot using S3, `upgrade to v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|:white_check_mark:|
|3|Setup Rancher `v2.7.2`, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, upgrade Rancher to `v2.7-head`, take an etcd snapshot,  `upgrade to v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|:white_check_mark:|
|4|Setup Rancher `v2.7.2`, provision a downstream RKE2 `v1.24.x+rke2r1` cluster, upgrade Rancher to `v2.7-head`, take an etcd snapshot using S3,  `upgrade to v1.25.x+rke2r1`, restore the snapshot and validate that all workloads come up as `Active`|:white_check_mark:|

**VALIDATION STEPS**

**Scenario 1**
1. As a standard user, provisioned a downstream RKE2 `v1.24.13+rke2r1` node driver cluster.
     - Validated that the cluster and nodes came up as `Active`.
2. Took an etcd snapshot of the cluster.
3. Upgraded the cluster to `v1.25.9+rke2r1`; validated that the cluster and nodes are `Active` and have had their K8s version upgraded.
4. Restored the snapshot with option cluster configuration, Kubernetes version and etcd.
     - Validated that the cluster and nodes are back at `v1.24.13+rke2r1` and they are `Active`.
5. Validated that all workloads are in a state of `Active`.
________
**Scenario 2**
1. Repeated scenario 1, but instead of taking a local etcd snapshot, I utilized an S3 bucket as my snapshot location.
     - Validated in scenario 1 step 2 that the backup appears in Amazon S3 under my respective bucket.
______
**Scenario 3**
1. Setup Rancher `v2.7.2` and logged in as a standard user.
2. Provisioned a downstream RKE2 `v1.24.13+rke2r1` node driver cluster.
     - Validated that the cluster and nodes came up as `Active`.
3. Upgraded Rancher to `v2.7-head`; validated that the cluster or nodes did not auto-upgrade after upgrade completion.
4. Took an etcd snapshot of the cluster.
5. Upgraded the cluster to `v1.25.9+rke2r1`; validated that the cluster and nodes are `Active` and have had their K8s version upgraded.
6. Restored the snapshot with option cluster configuration, Kubernetes version and etcd.
     - Validated that the cluster and nodes are back at `v1.24.13+rke2r1` and they are `Active`.
7. Validated that all workloads are in a state of `Active`.
_______
**Scenario 4**
1. Repeated scenario 3, but instead of taking a local etcd snapshot at step 4, I utilized an S3 bucket as my snapshot location.
     - Validated that the backup appears in Amazon S3 under my respective bucket.
